money-core-scala/src/main/scala/com/comcast/money/core/CoreSpan.scala

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2012-2015 Comcast Cable Communications Management, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.comcast.money.core
+
+import java.lang.Long
+
+import com.comcast.money.api._
+
+import scala.collection.JavaConversions._
+import scala.collection.concurrent.TrieMap
+
+/**
+ * A mutable implementation of the Span that also includes SpanInfo
+ *
+ * @param id The [[SpanId]]
+ * @param name The name of the span
+ * @param handler The [[SpanHandler]] responsible for processing the span once it is stopped
+ */
+case class CoreSpan(
+    id: SpanId,
+    name: String,
+    handler: SpanHandler
+) extends Span {
+
+  private var startTime: Long = 0L
+  private var startInstant: Long = 0L
+  private var endTime: Long = 0L
+  private var endInstant: Long = 0L
+  private var success: java.lang.Boolean = true
+
+  // use concurrent maps
+  private val timers = new TrieMap[String, Long]()
+  private val noted = new TrieMap[String, Note[_]]()
+
+  def start(): Unit = {
+    startTime = System.currentTimeMillis
+    startInstant = System.nanoTime
+  }
+
+  def stop(): Unit = stop(true)
+
+  def stop(result: java.lang.Boolean): Unit = {
+    endTime = System.currentTimeMillis
+    endInstant = System.nanoTime
+    this.success = result
+
+    handler.handle(this)
+  }
+
+  def stopTimer(timerKey: String): Unit =
+    timers.remove(timerKey) foreach { timerStartInstant =>
+      record(Note.of(timerKey, System.nanoTime - timerStartInstant))
+    }
+
+  def record(note: Note[_]): Unit = noted += note.name -> note
+
+  def startTimer(timerKey: String): Unit = timers += timerKey -> System.nanoTime
+
+  def info(): SpanInfo =
+    CoreSpanInfo(
+      id,
+      name,
+      startTime,
+      startInstant,
+      endTime,
+      endInstant,
+      endInstant - startInstant,
+      success,
+      noted.toMap[String, Note[_]]
+    )
```

Comment:
right, it would definitely be weird, a negative number in that case.

Seems like it would be better to have the duration be the time since the startInstant, if there is no endInstant use System.nanoTime instead, right?